blog-posts/event-sourcing-serveless/event-sourcing-serveless.md

```diff
@@ -0,0 +1,171 @@
+
+## Command
+
+We need to fuel this event store with some events. The command, which is the C of CQRS interface, is in charge of this. It's basically a handler that writes an event into the event store when triggered.
+
+Building our architecture we want something to accept HTTP request, so an API Gateway HTTP does the job. Whenever you pay 20$ with your card, a POST request is made to the endpoint <code>account/{accountId}/<b>debit</b></code> with a body `{ amount: 20 }`. Then this triggers a lambda function dedicated to writing **debit events** in the event store, and in this case it writes the event `{ 'accountId': 'abcd1234-user2', 'version': 12, 'eventType': 'Debit', 'amount': 20, 'timestamp': '2022-07-25T20:13Z'}`. According to the same principle, **credit events** are written into the event store.
+
+We now have our writing side with two routes, one for the debit transactions and one for credit transactions. Each route triggers a lambda that writes the event into the event store.
+
+<img src="./assets/archi-step-2-command.png" style="display: block; margin: auto" height="300"/>
+
+But we face an issue here. What happens if you want to buy blue jeans for 100$ but you only have 50$ and considering that we don't accept overdraft? We need **a way to know the current balance of your account** to implement some **business logic** to avoid unauthorized transactions. The computation of the current state of the account in question is called the aggregate.
+
+## Aggregate
+
+You need to remember that the source of truth of your application is the event store. Therefore when you want to make business verifications you have to do it on the event store and not on the projection which is only the read model.
+
+Coming back to the purchase of your 100$ blue jeans, how do we know that you only have 50$ left and that you can't perform this debit transaction only based on former transactions. That's where we build a tool named aggregate to compute the current balance. Before creating an event in the command, we query on the partition key with your id and we get a sorted list of your events based on the version of each one. Then we simply replay every debit and credit event in a reducer and output the total balance of the account. Unfortunately the computation of the balance reveals that you can't afford these blue jeans and we are able to send an error back.
+
+<img src="./assets/aggregate-schema.png" style="display: block; margin: auto" height="150"/>
+
+Likewise, we might need some business verification for the credit transaction. Thus our architecture should include a feedback loop between the command and the event store.
+
+<img src="./assets/archi-step-3-aggregate.png" style="display: block; margin: auto" height="300"/>
+
+At this moment we are able to write consistent data in our source of truth. But we don't serve data from the event store directly. We will implement database for read models.
```

Comment:
On comprend pas forcément cette phrase `We will implement database for read models.`
Tu veux dire qu'il y a une db séparée pour créer des read models?